### DISCUSSION TITLE
Postal installed in LXD/LXC container.... docker container "lost" after copying the LXD container from server to laptop.

### DISCUSSION THREAD

**Discussion by JohnHammell** [postalserver/postal#3186] [Help with using Postal]
My first Postal install, which was fairly recent, runs inside an LXD container. Everything seems to be running just right. I never have problems with restarting the LXD container as whenever I do Postal is right back up and running exactly as before the restart.

I'm running into a problem with copying the LXD container from my server to my laptop so I can do some local work on it, basically poke around a bit, etc, on a dev copy of the LXD container with Postal.

However, after I copy the LXD container with Postal none of docker containers are running. They all fail to start back up. Doing this same thing on my server to the original LXD container, restarting it, does not cause the Postal docker containers to stop running.

I've been able to get the web ui, smtp & worker postal docker containers back up pretty easily, just by running: postal start

The postal-mariadb container, though, kinda looks as though it's "lost." The folder for postal-mariadb is in /var/lib/docker, but when I try to start it I get a message that it is not found:
>root@lxd-container-name:~# docker start postal-mariadb
>Error response from daemon: No such container: postal-mariadb
>Error: failed to start containers: postal-mariadb

Any idea what is going on with this?

Literally, all I did was stopped the LXD container with the actively running Postal installation, took an LXD snapshot of the container (which is basically as ZFS snapshot as far as I know), then downloaded that snapshot to my laptop. Then I start that LXD container on my laptop and the issues I described arise. Not really sure why Postal does not start back up on its own but the main issue is the "lost" postal-mariadb docker container.

"lost" is in quotes because I really don't know. Maybe I'm doing something wrong here or maybe there's some command I'm not aware of which needs to be used to get postal-mariadb started up again.

**Comment by willpower232** [comment]
I don't really know much about LXD but I do know that maria and mysql can be a bit funky with copying data around depending on the table type.

I'm also presuming you are running it with a data volume to preserve its contents so it might be worth stopping the container before the snapshot, to ensure the data is definitely stored locally. If that doesn't work, presumably you could mysqldump it.

**Reply by JohnHammell** [reply]
LXD/LXC is basically just a virtual machine, but more efficient. LXC is what Docker was originally based on. LXD/LXC can run a full Linux OS, which is how I use it for hosting Postal and other software on my server (Postal in an LXC container/VM which is dedicated solely to Postal's install). Lots of conveniences in using it and it integrates very nicely with the ZFS filesystem, so nice optimizations there as well.

Copying the LXD/LXC VM container would be similar to simply imaging a physical hard drive. I would expect everything that's copied to be the same, but for some reason it's not. It could be that LXD/LXC is doing something to the networking part, external to the VM container, but not sure. I'll open a support request with the LXD project about this as well.

In regards to postal-mariadb, I installed that exactly as per the Postal installation docs prerequisites page at https://docs.postalserver.io/getting-started/prerequisites

```
docker run -d \
   --name postal-mariadb \
   -p 127.0.0.1:3306:3306 \
   --restart always \
   -e MARIADB_DATABASE=postal \
   -e MARIADB_ROOT_PASSWORD=postal \
   mariadb
```

The only deviation I made to the Postal install instructions was to use a different root password. Other than that, exactly the same as in the docs. Also, this postal-mariadb I installed into the same virtual machine / container as Postal. In addition, the VM/container is an Ubuntu 24.04 server, in case that makes any difference.

**Reply by willpower232** [reply]
Ah interesting, it doesn't mention using a volume so I'm going to say that the docker container doesn't keep its data.

If it says no such container, what docker containers does the restored LXD container list?

**Reply by JohnHammell** [reply]
The postal containers (smtp, web & worker) seem to function normally when the LXD container is copied to my laptop from the server, but the postal-mariadb container does not & ends up "missing."

Initially none are listed with 'docker ps -a' but 'postal start' gets the postal containers going, other than postal-mariadb container.

Haven't figured out why yet. The postal-mariadb docker container works just fine and doesn't lose anything when I start/stop & restart the container on the server. I'll be moving that installation over to a non-docker install of mariadb soon, though, just to ensure I can get a container backup running again if I ever need to.

**Reply by willpower232** [reply]
> Initially none are listed with 'docker ps -a' 

Aha well there you go, taking a snapshot of the container does not capture any running containers, not just the database. Its just more annoying with the database as that needs a manual start unless you add it to Postals docker-compose.yml (although you probably want to backup the database more directly before you mess around with it too much as there isn't a volume as previously mentioned)

**Reply by JohnHammell** [reply]
The strange part is that LXD/LXC uses ZFS snapshots (when ZFS is the filesystem as I have) and ZFS snapshots do not discriminate data. ZFS snapshots are simply a 100% snapshot of data at that specific point in time. I suppose it would be similar to making an image copy of a hard drive. After taking the lxc snapshot it can be seen in zfs list to view zfs snapshots.

I'm pretty sure I tried doing this taking the snapshot while the container is both offline & online with the same result. I'll test again, though. I'll post an update after doing that.

**Reply by JohnHammell** [reply]
Tested again. 2 snapshots, one with the lxd/lxc container vm turned off and the other with it running. Both had the same result:

"postal status" returned no containers running.

"postal start" re-downloads/pulls the postal containers and gets them going again.

"docker start postal-mariadb" returns "Error response from daemon: No such container: postal-mariadb"
